### ISSUE TITLE
Can't pass Environment variables from docker-compose.yml to Dockerfile

### ISSUE THREAD

**Issue by MatthewBooth** [docker/compose#5600]
This is a huge oversight. I'm not sure when the use-case would be that you want some variables within your project that build cannot access.

I'm running Laravel, which of course uses .env files. docker-composer passes these along nicely and the system runs.

Until I want to build an image and then composer (php) cannot generate autoload files because it can't find the .env file and therefore can't build an image.

TL;DR
My project needs some .env values and the Dockerfile has no way of knowing about them.

**Comment by shin-** [comment]
Compose is not a build tool, and has no vocation to become one. That being said, it should be simple enough to pass those variables you need as build args and declare them as `ENV` variables in your `Dockerfile`, like so:

*docker-compose.yml*
```yaml
version: '2.3'
services:
  foobar:
    build:
      context: .
      args:
        FOO: ${FOO}
        BAR: ${BAR}
```

*Dockerfile*
```dockerfile
FROM busybox:latest
ARG FOO
ARG BAR
ENV FOO $FOO
ENV BAR $BAR
...
```

**Comment by MatthewBooth** [comment]
Yes, I was using a Dockerfile for the build and had "build" and "context" within the compose file.

I appear to be missing this bit

```
ENV FOO $FOO
ENV BAR $BAR
```

I'll give that a shot tomorrow. The docs are not really clear on this and I cannot imagine my case is an edge case.

**Comment by landsman** [comment]
This is annoying...

**Comment by Resch-Said** [comment]
Some people like me are stupid and need clear names, so I will just make it a little more clear for myself and others.
docker-compose.yml
```
args:
         TOMCAT_ADMIN_PASSWORD: ${TOMCAT_PASSWORD}
```

Dockerfile
```
ARG TOMCAT_ADMIN_PASSWORD
ENV TOMCAT_ADMIN_PASSWORD $TOMCAT_ADMIN_PASSWORD
```

So, Compose passes the args to the Dockerfile
the `ARG TOMCAT_ADMIN_PASSWORD` contains `TOMCAT_PASSWORD`
and then I'm passing `ARG TOMCAT_ADMIN_PASSWORD` to `ENV TOMCAT_ADMIN_PASSWORD`

**Comment by lighterlightbulb** [comment]
This is not a practical solution whatsoever if you have a big environment file. This makes development tedious and unnecessarily annoying where if you can just read environment variables passed from the compose file. Hard-coding variables is never good, because if you would then want to add more environment variables you have to edit multiple files.

**Comment by dhvcc** [comment]
I guess the only way not to hardcode variables is to source your `.env` before running `docker-compose` like

```bash
. .env && docker-compose up --build -d
```

Keep in mind that `.` and `source` are the same command, yet I had problems with `source` in `Makefile`, so I stick with `.`

Also consider using `Makefile` for cleaner interface.

**Comment by mehrdad-shokri** [comment]
as mentioned by @Downvote-Hunter, just pass your args under build directive, then define them in Dockerfile by ``ARG ARG_NAME`` from then on you can access it by ``$ARG_NAME``

**Comment by lightbulblighter** [comment]
Again, that re-introduces the problem of hard-coding your environment variable names.

One thing I've noticed is that simply adding `env_file: "./.env"` to your docker-compose files sends the environment variables to the image, and it can also be accessed by the image. They can't be accessed by the Dockerfile directly, but the image can. You can use an entrypoint script to sort of mitigate any lost functionality.

**Comment by Edmartt** [comment]
hardcoded values works inside docker-compose, but not the same with env vars from .env. I f I exec a shell I can see my env vars exists but my app not working properly

**Comment by edjchapman** [comment]
> Compose is not a build tool, and has no vocation to become one. That being said, it should be simple enough to pass those variables you need as build args and declare them as `ENV` variables in your `Dockerfile`, like so:
> 
> _docker-compose.yml_
> 
> ```yaml
> version: '2.3'
> services:
>   foobar:
>     build:
>       context: .
>       args:
>         FOO: ${FOO}
>         BAR: ${BAR}
> ```
> 
> _Dockerfile_
> 
> ```dockerfile
> FROM busybox:latest
> ARG FOO
> ARG BAR
> ENV FOO $FOO
> ENV BAR $BAR
> ...
> ```

"Compose is not a build tool" -  yes ok, I don't think that's the issue here.  I think the issue and confusion here is that the the same env is not available at build time as it is at run time.  Specifying individual environment variables is problematic if you have lots of them.

**Comment by Booplicate** [comment]
It is a mystery why this process needs to be painful. First rule to learn: DRY, yet we have to write down every env var again?

> Compose is not a build tool, and has no vocation to become one

This is why it can build, then?

**Comment by dhvcc** [comment]
> It is a mystery why this process needs to be painful. First rule to learn: DRY, yet we have to write down every env var again?
> 
> > Compose is not a build tool, and has no vocation to become one
> 
> This is why it can build, then?

Having a way to pass build args, but not envs, while having a way of providing envs to the container. Also, a lot of repetition when you're providing through `environment:`

**Comment by scottillogical** [comment]
I would use "docker-compose run" instead of build and not have this issue, but with run, you can't define a multi-stage build target.   So if you want to run some code in a specific stage, you must use build (unless we add target to docker-compose run)

**Comment by Splines** [comment]
> ```shell
> . .env && docker-compose up --build -d
> ```

@dhvcc Did this work for you? For me, it doesn't. I have a `docker.env` file that looks like this:

```env
# docker.env
FANCY=5
```

Then I do:
```sh
# A .sh script
set -a # enable allexport (env variable) shell option
. ./docker.env
echo $FANCY # this works fine and prints "5"
docker compose build -d --pull # does not work
set +a # disable allexport (env variable) shell option
```

For testing purposes, I've included the following in the dockerfile:
```sh
# Dockerfile
RUN echo $FANCY
```
but it prints nothing (an empty line) when executed.

**Comment by dhvcc** [comment]
Hey, @Splines I doubt I'd know what happened at that time, it was quite a while ago. My best suggestion would be to use build args, `ARG` in `Dockerfile` and `build: ... args:` in compose. You can use access some environment variables in compose configs, like it usually auto loads `.env` files. See screenshot

<img width="427" alt="image" src="https://github.com/docker/compose/assets/18076967/e438c955-3c98-4b64-9218-593f2e958a0c">

To avoid repeating some of the variables you send you might want to look at [yaml anchors](https://support.atlassian.com/bitbucket-cloud/docs/yaml-anchors/)

Hope this helps

**Comment by ernstki** [comment]
> Keep in mind that `.` and `source` are the same command, yet I had problems with `source` in `Makefile`, so I stick with `.`

@dhvcc, great tip thanks!

In Bash, `source` and `.` _are_ the same command; in Z shell, they're _nearly_ the same. However, GNU Make uses `/bin/sh` [unless otherwise specified](https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html), which [does not have a `source` command](https://heirloom.sourceforge.net/sh/sh.1.html#14).

You can prove this to yourself with:

```console
$ echo $SHELL
/bin/bash

$ make -f <(printf 'test:\n\techo $(SHELL)\n')
echo /bin/sh
/bin/sh
```

**Comment by dhvcc** [comment]
> > Keep in mind that `.` and `source` are the same command, yet I had problems with `source` in `Makefile`, so I stick with `.`
> 
> @dhvcc, great tip thanks!
> 
> In Bash, `source` and `.` _are_ the same command; in Z shell, they're _nearly_ the same. However, GNU Make uses `/bin/sh` [unless otherwise specified](https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html), which [does not have a `source` command](https://heirloom.sourceforge.net/sh/sh.1.html#14).
> 
> You can prove this to yourself with:
> 
> ```console
> $ echo $SHELL
> /bin/bash
> 
> $ make -f <(printf 'test:\n\techo $(SHELL)\n')
> echo /bin/sh
> /bin/sh
> ```

> Yep, that's a more detailed answer. AFAIK `make` by default uses SH unless you [tell it to switch](https://www.gnu.org/software/make/manual/html_node/Choosing-the-Shell.html) (using `, so there were problems with using `source`.

Good to have more context, thanks!

**Comment by Splines** [comment]
> Hey, @Splines I doubt I'd know what happened at that time, it was quite a while ago. My best suggestion would be to use build args, [...]

@dhvcc, thanks for your suggestions! In the end, I filed in a [StackOverflow question](https://stackoverflow.com/q/78098380/9655481) tailored specifically for my use case ;)
